### PR TITLE
Dark theme

### PR DESCRIPTION
**Pull request by htwyford**
Addresses https://github.com/mozilla/side-view/issues/185. The onboarding in the sidebar and doorhanger now take on the dark theme if it's applied in Firefox. 

One outstanding issue is that Firefox broadcasts a switch from the default to the dark theme, but not from the the dark theme to the default. This means the sidebar changes when the dark theme is applied but doesn't return to default without a browser restart. I figured this is a smallish edge case: a user would have to switch from default to dark while the onboarding window is open. This issue doesn't cause any breaking changes. Doesn't seem like it can be resolved until https://bugzilla.mozilla.org/show_bug.cgi?id=1435216 is addressed.

Alternatively, we could disable the listener. This would allow us to remove the `management` permission. The user would have to close and reopen the sidebar after switching to dark theme to see the changes.

### CODE REVIEW

File: addon/theming.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
+var updateBackground = function(el, color) {
```
Comment (ericawright) [inline addon/theming.js]: use `let` and `const` instead of `var`

File: addon/popup.js [hunk 2]
```diff
@@ -199,6 +199,14 @@ async function init() {
     element("#private-warning").style.display = "";
   }
   updateHome();
+
+  handleThemes(isPopup = true); 
+  
+  // // watch for switch to dark theme.
+  // // BUG: No event transmitted when switching back to default
```
Comment (ericawright) [inline addon/popup.js]: remove this comment, since we aren't using themes other than dark/light

**Comment by ericawright** [comment]
Instead of using `isPopup`, can't we just have one listener and update both the popup and the sidebar when triggered?

**Comment by htwyford** [comment]
Regarding `isPopup`, the sidebar and popup are blind to each other, and the popup is only updated when it's opened. If you attempt to update the sidebar variables when the popup needs to be drawn, there's an error and both fail, and vice versa.

File: addon/theming.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        updateColor(tabs[i], textcolor);
+      }
+    }
+
+    // pseudoclasses must be updated by changing stylesheet
+    let css = "button.tab:hover{ background-color: " +
+      (theme.colors.popup_highlight || "#ededf0") + "!important }";
+    let style = document.createElement("style");
+
+    if (style.styleSheet) {
```
Comment (ericawright) [inline addon/theming.js]: does this ever evaluate to true?

File: addon/theming.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        updateColor(links[i], theme.colors.textcolor);
+    }
+  }
+}
+
+function updatePopupStyle() {
+  if (theme.colors) {
+    let panel = document.getElementById("panel");
+    let tabs = document.getElementsByClassName("tab");
+    let feedback = document.getElementsByClassName("feedback-button")[0];
```
Comment (ericawright) [inline addon/theming.js]: document.getQueryselector('feedback-button') will get you only the first element it finds with that selector

File: addon/theming.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+const updateBackground = function(el, color) {
+  el.style.background = color;
+};
+
+const updateColor = function(el, color) {
+  el.style.color = color;
+};
+
+function updateSidebarStyle() {
+  if (theme.colors) {
```
Comment (ericawright) [inline addon/theming.js]: can these styles be applied with css instead? select a parent element el.classList.add('dark-theme') then style all the children within it dark?
We tlaked about how the border and carat of the panel can only be changed with JS, but in all other cases css is better

**Comment by htwyford** [comment]
The selection of what elements to dark theme is more specific than "all children of root" to avoid writing too many exceptions in the CSS.

File: addon/sidebar.css [hunk 6]
```diff
@@ -147,3 +147,20 @@ body {
...
+
+.page.dark-theme {
+  color: #fff;
+  background: #4a4a4f;
+}
+
+.page.dark-theme button:hover,
+.page.dark-theme button:focus,
+.page.dark-theme .graphic__shadow {
+  background: #6d6d6f;
```
Comment (ericawright) [inline addon/sidebar.css]: I like the effect on the shadow, but it creates a strange highlighter effect on the links.

File: addon/sidebar.js [hunk 7]
```diff
@@ -2,6 +2,25 @@ function element(selector) {
   return document.querySelector(selector);
 }
 
+function applyDarkTheme() {
+  document.body.style.background = "#4a4a4f";
+  document.body.style.color = "#fff";
+  document.querySelector(".page").classList.add("dark-theme");
```
Comment (ericawright) [inline addon/sidebar.js]: looks like there's already a utility function written above on this page for this, we should use it.

**Comment by ianb** [comment]
Is there anything with `browser_style` that would help us here? (A property mentioned on [this page](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/manifest.json/browser_action))

There's a possibility we could detect the theme using browser_style and getComputedStyle(). (Note that I don't know if browser_style even respects the dark theme, and I haven't tested any of this, it's just a hunch.)

There's also a possibility that if we don't specify things browser_style already gives us, then the dark theme would magically be applied? (That seems unlikely to me.)

**Comment by htwyford** [comment]
`browser_style` doesn't seem to respect dark theme. Searching through chrome://browser/content/extension.css doesn't turn up any references to any of the dark theme colours (or the works "dark", "night", "firefox-compact-dark", etc., for that matter). 
I also tried removing specific references to `browser_style` as well, and it just defaults to the usual white look. I think something custom is needed, as least until `browser_style` starts broadcasting dark-theme changes.

**Comment by ericawright** [comment]
@htwyford this is ready right? @fzzzy or @ianb  want to review again?

**Comment by htwyford** [comment]
It's ready!

**Comment by fzzzy** [comment]
@htwyford This looks good, but it has some conflicts now. Can you resolve them? Thanks!

**Comment by fzzzy** [comment]
Nice!

**Pull request merged**